core/src/main/java/org/apache/druid/timeline/partition/OvershadowableManager.java

```diff
@@ -189,89 +247,359 @@ private void transitAtomicUpdateGroupState(AtomicUpdateGroup<T> atomicUpdateGrou
-      // than the given minorVersion.
-      final Short2ObjectSortedMap<AtomicUpdateGroup<T>> versionToGroup = current.getValue();
-      // Short2ObjectRBTreeMap.SubMap.short2ObjectEntrySet() implementation, especially size(), is not optimized.
-      // Note that size() is indirectly called in ArrayList.addAll() when ObjectSortedSet.toArray() is called.
-      // See AbstractObjectCollection.toArray().
-      // If you see performance degradation here, probably we need to improve the below line.
-      found.addAll(versionToGroup.subMap(versionToGroup.firstShortKey(), minorVersion).short2ObjectEntrySet());
-      current = stateMap.higherEntry(current.getKey());
-    }
-    return found;
+    return current;
   }
 
   /**
-   * Handles addition of the atomicUpdateGroup to the given state
+   * Determine the visible group after a new chunk is added.
    */
-  private void transitionStandbyGroupIfFull(AtomicUpdateGroup<T> aug, State stateOfAug)
+  private void determineVisibleGroupAfterAdd(AtomicUpdateGroup<T> aug, State stateOfAug)
   {
     if (stateOfAug == State.STANDBY) {
       // A standby atomicUpdateGroup becomes visible when its all segments are available.
       if (aug.isFull()) {
         // A visible atomicUpdateGroup becomes overshadowed when a fully available standby atomicUpdateGroup becomes
         // visible which overshadows the current visible one.
-        findOvershadowedBy(aug, State.VISIBLE)
-            .forEach(entry -> transitAtomicUpdateGroupState(entry.getValue(), State.VISIBLE, State.OVERSHADOWED));
+        replaceVisibleWith(
+            findOvershadowedBy(aug, State.VISIBLE),
+            State.OVERSHADOWED,
+            Collections.singletonList(aug),
+            State.STANDBY
+        );
         findOvershadowedBy(aug, State.STANDBY)
-            .forEach(entry -> transitAtomicUpdateGroupState(entry.getValue(), State.STANDBY, State.OVERSHADOWED));
-        transitAtomicUpdateGroupState(aug, State.STANDBY, State.VISIBLE);
+            .forEach(entry -> transitAtomicUpdateGroupState(entry, State.STANDBY, State.OVERSHADOWED));
+      } else {
+        // The given atomicUpdateGroup is in the standby state which means it's not overshadowed by the visible group.
+        // If the visible group is not fully available, then the new standby group should be visible since it has a
+        // higher minor version.
+        checkVisibleIsFullyAvailableAndMoveNewStandbyToVisible(aug, stateOfAug);
       }
+    } else if (stateOfAug == State.OVERSHADOWED) {
+      checkVisibleIsFullyAvailableAndTryToMoveOvershadowedToVisible(aug, stateOfAug);
     }
   }
 
-  private void addAtomicUpdateGroupWithState(AtomicUpdateGroup<T> aug, State state)
+  /**
+   * This method is called in {@link #determineVisibleGroupAfterAdd}. It checks the current visible group is
+   * fully available, and if not, moves the given group to the visible state.
+   */
+  private void checkVisibleIsFullyAvailableAndMoveNewStandbyToVisible(
```

Comment:
Suggest adding a comment about how this method assumes the standby group is not full. I forgot that there was an `aug.isFull()` check in `determineVisibleGroupAfterAdd` and was confused why this method didn't replace incomplete visible with the standby. Or maybe move the full standby handling into this method as well

Reply:
Good point. Moved the full standby handling into this method and updated its javadoc.